org.eclipse.xtext.xbase/src/org/eclipse/xtext/xbase/compiler/XbaseCompiler.java

```diff
@@ -1132,14 +1131,9 @@ protected void constructorCallToJavaExpression(final XConstructorCall expr, ITre
 	}
 
 	protected void appendConstructedTypeName(XConstructorCall constructorCall, ITreeAppendable typeAppendable) {
-		JvmDeclaredType type = constructorCall.getConstructor().getDeclaringType();
-		if (type instanceof JvmGenericType && ((JvmGenericType) type).isAnonymous()) {
-			typeAppendable.append(Iterables.getLast(type.getSuperTypes()).getType());
-		} else {
-			IResolvedTypes resolvedTypes = batchTypeResolver.resolveTypes(constructorCall);
-			LightweightTypeReference actualType = resolvedTypes.getActualType(constructorCall).getRawTypeReference();
-			typeAppendable.append(actualType);
-		}
+		IResolvedTypes resolvedTypes = batchTypeResolver.resolveTypes(constructorCall);
+		LightweightTypeReference actualType = resolvedTypes.getActualType(constructorCall).getRawTypeReference();
```

Comment:
Should this use getNamedType to obtain the correct type if it’s an anonymous class?

Reply:
@szarnekow This is done by org.eclipse.xtext.xbase.typesystem.references.LightweightTypeReferenceSerializer.doVisitParameterizedTypeReference(ParameterizedTypeReference)

```java
	@Override
	protected void doVisitParameterizedTypeReference(ParameterizedTypeReference reference) {
		if (reference.isAnonymous()) {
			reference.getNamedType().accept(this);
		} else {
			appender.append(reference.getType());
			if (!reference.getTypeArguments().isEmpty()) {
				appender.append("<");
				appendCommaSeparated(reference.getTypeArguments());
				appender.append(">");
			}
		}
	}
```

But I've just noted Xtend compilation tests fail: also the org.eclipse.xtext.xbase.compiler.XbaseCompiler.constructorCallToJavaExpression(XConstructorCall, ITreeAppendable) needs adjusting/simplification or it generates type arguments twice.

Reply:
@szarnekow the new commit https://github.com/eclipse/xtext/pull/2888/commits/fdad9842c53185582bc15547e96118b3cedf82e0 updates/simplified org.eclipse.xtext.xbase.compiler.XbaseCompiler.constructorCallToJavaExpression(XConstructorCall, ITreeAppendable) accordingly (at least locally Xtend tests are now green)

Reply:
@szarnekow it can be simplified even further https://github.com/eclipse/xtext/pull/2888/commits/28a2e11d74f3c671e6b18f0d24c7af99046b6f36

last message for this year: Happy new year :)